src/controller/mod.rs

```diff
@@ -78,56 +85,18 @@ impl WorkerStatus {
     }
 }
 
-
-enum ControlEvent {
-    WorkerCoordination(CoordinationMessage),
-    ExternalGet(String, String, Sender<String>),
-    ExternalPost(String, String, Sender<String>),
-}
-
 /// Used to construct a controller.
 pub struct ControllerBuilder {
-    sharding: Option<usize>,
-    domain_config: DomainConfig,
-    persistence: PersistenceParameters,
-    materializations: migrate::materialization::Materializations,
+    config: ControllerConfig,
     listen_addr: IpAddr,
-    heartbeat_every: Duration,
-    healthcheck_every: Duration,
-    nworkers: usize,
-    local_workers: usize,
-    internal_port: u16,
-    external_port: u16,
-    checktable_port: u16,
     log: slog::Logger,
 }
 impl Default for ControllerBuilder {
     fn default() -> Self {
-        let log = slog::Logger::root(slog::Discard, o!());
         Self {
-            #[cfg(test)]
-            sharding: Some(2),
-            #[cfg(not(test))]
-            sharding: None,
-            domain_config: DomainConfig {
-                concurrent_replays: 512,
-                replay_batch_timeout: time::Duration::from_millis(1),
-                replay_batch_size: 32,
-            },
-            persistence: Default::default(),
-            materializations: migrate::materialization::Materializations::new(&log),
+            config: ControllerConfig::default(),
             listen_addr: "127.0.0.1".parse().unwrap(),
```

Comment:
Should this come from an argument? `src/controller/main.rs` has a parameter `address` that gets this information from the command line, but it doesn't seem to be passed into the `ControllerBuilder` anywhere.

EDIT: There's a `set_listen_addr` method below, but the `main.rs` doesn't call it -- should it do so?

Reply:
Good catch

Reply:
Wait, yes it does: https://github.com/mit-pdos/distributary/blob/39a6b0a28cb84f1087068db229256a39eda65229/src/controller/main.rs#L36